core/service/test/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+			}
+			network.run_until_all_full(|_index, service| {
+				service.network().status().num_peers == NUM_NODES as usize - 1
+			});
+		}
+		temp.close().expect("Error removing temp dir");
+	}
+	{
+		info!("Checking linked topology");
+		let temp = TempDir::new("substrate-connectivity-test").expect("Error creating test dir");
+		{
+			let mut network = TestNet::<F>::new(&temp, spec, NUM_NODES as u32, 0, vec![], 30400);
+			let mut address = network.full_nodes[0].1.network().node_id().unwrap();
+			for (_, service) in network.full_nodes.iter().skip(1) {
+				service.network().add_reserved_peer(address.clone()).expect("Error adding reserved peer");
+				address = service.network().node_id().unwrap();
+			}
+			network.run_until_all_full(|_index, service| {
+				service.network().status().num_peers == NUM_NODES as usize - 1
+			});
+		}
+		temp.close().expect("Error removing temp dir");
+	}
+}
+
+pub fn sync<F, B>(spec: FactoryChainSpec<F>, block_factory: B)
+where
+	F: ServiceFactory,
+	B: Fn(&F::FullService) -> (JustifiedHeader<F::Block>, Option<Vec<FactoryExtrinsic<F>>>),
+{
+	const NUM_NODES: u32 = 10;
+	const NUM_BLOCKS: usize = 512;
+	info!("Checking block sync");
+	let temp = TempDir::new("substrate-sync-test").expect("Error creating test dir");
+	let mut network = TestNet::<F>::new(&temp, spec.clone(), NUM_NODES, 0, vec![], 30500);
+	let first_address = {
+		let first_service = &network.full_nodes[0].1;
+		for i in 0 .. NUM_BLOCKS {
+			if i % 128 == 0 {
+				info!("Generating #{}", i);
+			}
+			let (header, body) = block_factory(&first_service);
+			first_service.client().import_block(BlockOrigin::File, header, body, true).expect("Error importing test block");
+		}
+		first_service.network().node_id().unwrap()
+	};
+	info!("Running sync");
+	for (_, service) in network.full_nodes.iter().skip(1) {
+		service.network().add_reserved_peer(first_address.clone()).expect("Error adding reserved peer");
+	}
+	network.run_until_all_full(|_index, service| {
+		service.client().info().unwrap().chain.best_number == As::sa(NUM_BLOCKS as u64)
+	});
+}
+
+pub fn consensus<F>(spec: FactoryChainSpec<F>, authorities: Vec<String>)
+where
+	F: ServiceFactory,
+{
+	const NUM_NODES: u32 = 10;
+	const NUM_BLOCKS: u64 = 200;
+	info!("Checking consensus");
+	let temp = TempDir::new("substrate-conensus-test").expect("Error creating test dir");
+	let mut network = TestNet::<F>::new(&temp, spec.clone(), NUM_NODES, 0, authorities, 30600);
+	let first_address = network.authority_nodes[0].1.network().node_id().unwrap();
+	for (_, service) in network.full_nodes.iter() {
+		service.network().add_reserved_peer(first_address.clone()).expect("Error adding reserved peer");
+	}
+	for (_, service) in network.authority_nodes.iter().skip(1) {
+		service.network().add_reserved_peer(first_address.clone()).expect("Error adding reserved peer");
+	}
+	network.run_until_all_full(|_index, service| {
+		service.client().info().unwrap().chain.best_number >= As::sa(NUM_BLOCKS)
```

Comment:
perhaps it should check the number of the highest finalized block -- this would indicate that consensus is really working in the case of e.g. aura + GRANDPA